src/ORM/ResultSetFactory.php

```diff
@@ -34,15 +35,24 @@ class ResultSetFactory
      * Constructor
      *
      * @param \Cake\ORM\Query\SelectQuery<T> $query Query from where results came.
-     * @param array $results Results array.
+     * @param iterable $results Results.
      * @return \Cake\ORM\ResultSet<array|\Cake\Datasource\EntityInterface>
      */
-    public function createResultSet(SelectQuery $query, array $results): ResultSet
+    public function createResultSet(SelectQuery $query, iterable $results): ResultSet
     {
         $data = $this->collectData($query);
 
-        foreach ($results as $i => $row) {
-            $results[$i] = $this->groupResult($row, $data);
+        if (is_array($results)) {
+            foreach ($results as $i => $row) {
+                $results[$i] = $this->groupResult($row, $data);
+            }
+
+            $results = SplFixedArray::fromArray($results);
+        } else {
+            $results = (new Collection($results))
+                ->map(function ($row) use ($data) {
+                    return $this->groupResult($row, $data);
+                });
```

Comment:
Hopefully users don't iterator over the result set multiple times. Should we unwrap this before passing to ResultSet()?

Reply:
Is the non-array path the unbuffered statement so we can't unwrap this? Should we check it's a statement?

Reply:
Yes a non array path means it's an buffered statement. But since the `Collection` can handle any iterable, I don't want to narrow this down unnecessarily using a `Statement` check.